### PR TITLE
[1817] Volatility-variant private "Inventor" pays too much on 6 train purchase

### LINKED ISSUE #8988: [1817] Volatility-variant private "Inventor" pays too much on 6 train purchase
**What's your Game ID?**
https://18xx.games/game/106175
https://18xx.games/game/99154

https://18xx.games/game/106175?action=1524
https://18xx.games/game/99154?action=1119

https://gist.github.com/benjaminxscott/d313d464aa2505f14915e475ffc20876

The inventor company should only pay out $60 for the first 6 train purchase/export. Instead it pays $60 and then makes another payment of $30.

### PR DESCRIPTION
**Pull request by benjaminxscott**
Fixes #8988

### Before clicking "Create"

- [x] Branch is derived from the latest `master`
- [x] Add the `pins` label if this change will break existing games
- [x] Code passes linter with `docker compose exec rack rubocop -a`
- [x] Tests pass cleanly with `docker compose exec rack rake`

### Implementation Notes

* **Explanation of Change**

- Sets the `rusted` flag on trains before `remove_train()` is called, as within that function it was not possible to check the `rusted` state otherwise. This behavior was causing a bug as described here https://github.com/tobymao/18xx/issues/8988#issuecomment-1467477512

-  Also Removes an un-necessary "2+ trains provide zero revenue" definition from the lookup hash, since the calling function returns zero anyway if the train type is not in the hash

* **Screenshots**
Screenshot 2023-03-14 12 06 21 AM

* **Any Assumptions / Hacks**

We assume that `remove_train` will operate correctly if the train is already marked as rusted

### CODE REVIEW

File: lib/engine/game/base.rb [hunk 1]
```diff
@@ -1950,9 +1950,9 @@ def remove_train(train)
       end
 
       def rust(train)
+        train.rusted = true
         remove_train(train)
         train.owner = nil
-        train.rusted = true
```
Comment (crericha) [inline lib/engine/game/base.rb:1955]: This is probably okay. Did you check all games that implement remove_train to verify this change doesn't mess them up?

File: lib/engine/game/g_1817/game.rb [hunk 2]
```diff
@@ -873,8 +873,9 @@ def rust(train)
         def remove_train(train)
           from_depot = train.from_depot?
           super
+
           return unless inventor_private&.owner&.corporation?
-          return if !from_depot || @depot.trains.find { |t| t.name == train.name } != train
+          return if train.rusted || !from_depot || @depot.trains.find { |t| t.name == train.name } != train
```
Comment (crericha) [inline lib/engine/game/g_1817/game.rb:878]: I don't remember what this logic is attempting to do and I think it is hard to follow. I think we should go a simpler way of removing the value from @payouts after it pays out. That way, it'll only pay out once, like it should, and we don't have to track any additional state.

To do this, move the log message  and the return unless inventor_private&.owner&.corporation? to inventor_payout  and handle everything there. Then the return if rusted, !depot, etc. logic can be removed.

File: lib/engine/game/base.rb (hunk 1 above)
Comment (benjaminxscott) [reply lib/engine/game/base.rb:1955]: I looked at '28 and NY which seem to be the only games that shadow this function, and `rusted` doesn't seem to be a flag that they are checking or influencing their logic

Comment (benjaminxscott) [reply lib/engine/game/base.rb:1955]: OK with keeping this change to base code even if the bugfix implementation doesn't require it?

File: lib/engine/game/g_1817/game.rb (hunk 2 above)
Comment (benjaminxscott) [reply lib/engine/game/g_1817/game.rb]: Ah, made a commit that should implement this approach, check it out [5db9212](https://github.com/tobymao/18xx/pull/9011/commits/5db92125ab6b40c5c2969d8556d378e7330e3aec)

File: lib/engine/game/base.rb (hunk 1 above)
Comment (crericha) [reply lib/engine/game/base.rb:1955]: remove_train is the method in question. It is overridden by 17, 22, 28, 77, and NY.

File: lib/engine/game/g_1817/game.rb (hunk 2 above)
Comment (crericha) [reply lib/engine/game/g_1817/game.rb]: Looks good, I think.

File: lib/engine/game/base.rb (hunk 1 above)
Comment (benjaminxscott) [reply lib/engine/game/base.rb:1955]: Ah yes I checked 22 as well which seemed to be fine

'17 is this one so it should work

77 I think is a different function? `def remove_trainless`

**Comment by benjaminxscott** [comment]
@crericha OK to merge?

**Comment by crericha** [comment]
Yeah, looks good to me.
